src/components/PaginatedVirtualList/usePaginatedVirtualList.ts

```diff
@@ -0,0 +1,152 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import { VirtuosoHandle } from "react-virtuoso";
+import usePrevious from "hooks/usePrevious";
+import { leaveBreadcrumb } from "utils/errorReporting";
+import { calculatePageSize, calculateStartingIndex } from "./utils";
+
+interface UsePaginatedVirtualListProps {
+  rowCount: number;
+  paginationThreshold: number;
+  paginationOffset: number;
+  ref: React.RefObject<Pick<VirtuosoHandle, "scrollToIndex">>;
+}
+
+const usePaginatedVirtualList = ({
+  rowCount,
+  paginationOffset,
+  paginationThreshold,
+  ref,
+}: UsePaginatedVirtualListProps) => {
+  const [currentPage, setCurrentPage] = useState(0);
+  const prevPage = usePrevious(currentPage);
+  const totalPageCount = Math.ceil(rowCount / paginationThreshold);
+
+  // isAutoScroll is used to prevent the list from automatically adjusting the scroll position
+  // when jumpToLine is called. This is necessary because the list will automatically scroll to the
+  // top of the page when the user scrolls to the next or previous page. We don't want to scroll to
+  // the top of the page when jumpToLine is called, so we use this flag to prevent that from happening.
+  const isScrollerScroll = useRef(true);
+
+  const offsetCompensation = currentPage > 0 ? paginationOffset : 0;
+
+  const pageSize = calculatePageSize({
+    maxPageSize: paginationThreshold,
+    totalItemCount: rowCount,
+    currentPage,
+  });
+
+  const scrollToNextPage = useCallback(() => {
+    const nextPage = currentPage + 1;
+    if (nextPage < totalPageCount) {
+      setCurrentPage(nextPage);
+    }
+  }, [currentPage, totalPageCount]);
+
+  const scrollToPrevPage = useCallback(() => {
+    const nextPage = currentPage - 1;
+    if (currentPage !== 0) {
+      setCurrentPage(nextPage);
+    }
+  }, [currentPage]);
+
+  // The following useEffect is necessary to scroll to the correct position when the user
```

Comment:
(comment placement is random, nothing to change) Mysterious, but we should keep an eye out for this final-line-disappearing-thing !

<img width="1019" alt="Screenshot 2023-04-19 at 2 13 36 PM" src="https://user-images.githubusercontent.com/47064971/233163718-8299b01c-c8d1-42b3-80b9-14f776ad9825.png">